doc/developer/design/2024_10_08_expression_cache.md

```diff
@@ -0,0 +1,138 @@
+# Expression Cache
+
+## The Problem
+
+Optimization is a slow process that is on the critical path for `environmentd` startup times. Some
+environments spend 30 seconds just on optimization. All optimization time spent in startup is
+experienced  downtime for users when `environmentd` restarts.
+
+## Success Criteria
+
+Startup spends less than 1 second optimizing expressions.
+
+## Solution Proposal
+
+The solution being proposed in this document is a cache of optimized expressions. During startup,
+`environmentd` will first look in the cache for optimized expressions and only compute a new
+expression if it isn't present in the cache. If enough expressions are cached and the cache is fast
+enough, then the time spent on this part of startup should be small.
+
+The cache will present similarly as a key-value value store where the key is a composite of
+
+  - deployment generation
+  - object global ID
+  - expression type (local MIR, global MIR, LIR, etc)
+
+The value will be a serialized version of the optimized expression. An `environmentd` process with
+deploy generation `n`, will never be expected to look at a serialized expression with a deploy
+generation `m` s.t. `n != m`. Therefore, there are no forwards or backwards compatibility needed on
+the serialized representation of expressions. The cache will also be made  durable so that it's
+available after a restart, at least within the same deployment generation.
+
+Upgrading an environment will look something like this:
+
+1. Start deploy generation `n` in read-only mode.
+2. Populate the expression cache for generation `n`.
+3. Start deploy generation `n` in read-write mode.
+4. Read optimized expressions from cache.
+
+Restarting an environment will look something like this:
+
+1. Start deploy generation `n` in read-write mode.
+2. Read optimized expressions from cache.
+
+### Cache API
+
+Below is the API that the cache will present. It may be further wrapped with typed methods that
+take care of serializing and deserializing bytes. Additionally, we probably don't need a trait when
+implementing.
+
+```Rust
+trait ExpressionCache {
+    /// Returns the `expression_type` of `global_id` that is currently deployed in a cluster. This
+    /// will not change in-between restarts as result of DDL, as long as `global_id` exists.
+    fn get_deployed_expression(&self, global_id: GlobalId, expression_type: ExpressionType) -> Option<Bytes>;
```

Comment:
I added some notes, the TLDR is that it's an existing requirement for `EXPLAIN`.

Reply:
I guess what I'm confused about is why it needs to be part of the cache. Shouldn't the existing machinery that caches this in the catalog just continue to work as it already does?

Reply:
> Shouldn't the existing machinery that caches this in the catalog just continue to work as it already does?

Yes, but populating the existing cache is a huge contributor to the slowdown during startup. The main goal of this work is to remove that slowdown. For our largest customer it takes ~15 seconds. So we can keep all the existing machinery, but we'd want to populate it initially with our durable cache.

Does that answer your question? If not, I think I might be confused about what you're asking.

Reply:
Ah, sorry, what I don't understand is why `trait ExpressionCache` needs to be the thing that keeps track of what got deployed to the cluster. It seems to me like the catalog can bootstrap the existing `CatalogPlans` struct with what's in the cache (via `get_durable_expressions`), and then never read from the cache again in the steady state. (It'll keep _writing_ to the cache, of course).

Is your vision that we delete `CatalogPlans` and the various `optimized_expr` fields from the catalog objects, and then have explain sequencing interact with the expression cache to get the expressions?

I think that works too—just a question of which component is responsible for what. My sense is it's a smaller/easier change if the `ExpressionCache` is only concerned with durability, but maybe it results in a system that's harder to reason about because there are two separate places expressions are cached. Not sure! Haven't much thought about it.

Reply:
> Not sure! Haven't much thought about it.

Reading Alexander's original design document, though, it seems like it's tricky to have a single thing responsible for tracking both the local MIR and the global MIR, because the local MIR needs to be presented to the catalog in the txn that adds the object, and the global MIR is only computable after the thing is present in the catalog.

Reply:
Ah OK, I understand now.

> Is your vision that we delete `CatalogPlans` and the various `optimized_expr` fields from the catalog objects, and then have explain sequencing interact with the expression cache to get the expressions?

Almost. My original vision was that we delete `CatalogPlans`, but keep the various `optimized_expr` fields in the catalog objects, and then have explain sequencing interact with the expression cache to get the expressions.

I'll have to think about this a bit more to see what's the best approach.

Reply:
> Almost. My original vision was that we delete `CatalogPlans`, but keep the various `optimized_expr` fields in the catalog objects, and then have explain sequencing interact with the expression cache to get the expressions.

Ah, gotcha. What seems a little scary about that is that, IIUC, you now have the local MIR stored in memory in two places, and the possibility for them to get out of sync.

> FWIW, I'm also worried that this will make storing all expressions for a single `GlobalId` in a single blob a bit tricky.

Indeed. If you instead relegate the expression cache to just being used for the durableization of the expressions I think it becomes workable, though. The cache is allowed to lag behind the real catalog arbitrarily, which means you don't need to populate it until after the catalog _and_ `CatalogPlans` have been updated. And you're only allowed to use it during bootstrap, after calling `reconcile`.

If you go down this route, I think you could even change the API of the cache so that you can only read via a `reconcile_and_read` method that first reconciles the cache and then returns you all the entries in the cache. That would essentially guarantee that the cache was only used during bootstrapping.

Reply:
Hmm, yeah, I think I like that direction. It simplifies a lot of stuff. I'll update the API to reflect this.

Reply:
Done in https://github.com/MaterializeInc/materialize/pull/29908/commits/57e8d648b2acd4707b9acfed5997055a066a5020. We could even take it a step further and combine `new` with `reconcile_and_open`.

Reply:
We could even take it one more step further with a method like 
```Rust
reconcile_and_open_and_maybe_cleanup_all_prior_deploy_generations(deploy_generation: u64, current_ids: &BTreeSet<GlobalId>, read_only: bool) -> (Self, Vec<(GlobalId, Expressions)>)
```
, but that might be crazy.

Reply:
I don’t think that’s crazy! Call that method `open` and have it take and return structs and it looks super solid.

It does occur to me that in this world you probably want a `struct ExpressionCache` rather than a `trait ExpressionCache` since I can’t imagine we’ll want to implement a trait like this twice. Feels like there’s a deeper trait (`trait ExpressionCacheStorage`) that we could pull out if we need/want to evaluate both a persist and file storage backend for this thing.